flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java

```diff
@@ -144,6 +188,17 @@ private void maybeRescale() {
         }
     }
 
+    private Duration timeSinceLastRescale() {
+        return Duration.between(lastRescale, Instant.now());
+    }
+
+    private void rescaleWhenCooldownPeriodIsOver() {
+        context.runIfConditionOrSchedule(
+                timeSinceLastRescale().compareTo(scalingIntervalMin) > 0,
+                this::maybeRescale,
+                scalingIntervalMin); // reset cooldown period
```

Comment:
Schedule the `maybeRescale` after  `scalingIntervalMin` might not make sense.

For example, `scalingIntervalMin` is 30 seconds, `lastRescale` is `08:01:05`, and now is `08:01:10`.

From your code, the `maybeRescale` is called at `08:01:40`, however it should be called at `08:01:35`, right?

Reply:
That is right. My first implementation scheduled ad 08:01:35 but @dmvk suggested [here](https://lists.apache.org/thread/m2w2xzfjpxlw63j0k7tfxfgs0rshhwwr) that we keep the min interval low but we restart the cooldown period each time. That is what is expressed in the FLIP